### DISCUSSION TITLE
RFC - More modular way to write getStaticPaths/getStaticProps/getSSP

### DISCUSSION THREAD

**Discussion by Vadorequest** [UnlyEd/next-right-now#348] [Show and tell]
PR: https://github.com/UnlyEd/next-right-now/pull/346 and #347 

Very basically, I've changed the way to write the 3 main functions for generating Next.js pages.

The new way allows for a more modular design and better reusability of the source code.

#### Before:
```tsx
const getCoreLayoutStaticPaths: GetStaticPaths<CommonServerSideParams> = async (context: GetStaticPathsContext): Promise<StaticPathsOutput> => {
  // Implementation
}
```

#### After:
```tsx
export const getCoreLayoutStaticPaths: GetCoreLayoutStaticPaths = (options?: GetCoreLayoutStaticPathsOptions) => {
  const {
    fallback = false,
  } = options || {};

  const getStaticPaths: GetStaticPaths<CommonServerSideParams> = async (context: GetStaticPathsContext): Promise<StaticPathsOutput> => {
  // Implementation
  };

  return getStaticPaths;
};
```

Now, the **call site** (usually, the page itself) has more power, it can re-use part of the logic while providing `options` to disable some behaviour.

```tsx
// In some page
export const getStaticPaths: GetStaticPaths<CommonServerSideParams> = getDemoLayoutStaticPaths();

// In 404 page
export const getStaticProps: GetStaticProps<SSGPageProps, CommonServerSideParams> = getPublicLayoutStaticProps({ enable404Redirect: false });
```

Share your thoughts! 😄

**Comment by samuelcastro** [comment]
> export const getCoreLayoutStaticPaths: GetCoreLayoutStaticPaths = (options?: GetCoreLayoutStaticPathsOptions) => {
>   const {
>     fallback = false,
>   } = options || {};
> 
>   const getStaticProps: GetStaticProps<SSGPageProps, CommonServerSideParams> = async (props: StaticPropsInput): Promise<GetStaticPropsResult<SSGPageProps>> => {
>   // Implementation
>   };
> 
>   return getStaticProps;
> };

This does not look right, `getCoreLayoutStaticPaths` is returning `getStaticProps` is this what you meant?

**Reply by Vadorequest** [reply]
Yeah, I mixed the function names in the example.

**Reply by Vadorequest** [reply]
I fixed it, thanks!
